File: tests/api/api_objects/base_api.py


class BaseApi:
    """Base class for all API objects"""

    def __init__(self, base_url, session = None):
        self.base_url = base_url.rstrip('/')
        self.session = session

    def set_session(self, session):
        self.session = session
        return self

    def _get(self, endpoint, **kwargs):
        url = f"{self.base_url}{endpoint}"
        return self.session.get(url, **kwargs)
    
    def _post(self, endpoint, data, **kwargs):
        url = f"{self.base_url}{endpoint}"
        return self.session.post(url, data=data, **kwargs)
    
    def _delete(self, endpoint, **kwargs):
        url = f"{self.base_url}{endpoint}"
        return self.session.delete(url, **kwargs)
    
    def get_csrf_token(self, endpoint):
        """Extract csrf token from HTML using regex"""
        html_content = self.session.get(f"{self.base_url}{endpoint}").text
        import re
        match = re.search(r'name="csrfmiddlewaretoken" value="([^"]+)"', html_content)
        return match.group(1) if match else None

File: tests/load_tests/pages/base_api.js
import http from 'k6/http'
import { urls } from '../config/urls.js'


export class BaseApi {
    constructor(base_url=urls.base_url) {
        this.base_url = base_url;
        this.session = http;
    }

    _get(endpoint) {
        let url = `${this.base_url}${endpoint}`;
        return this.session.get(url)
    }

    getCsrfToken(endpoint) {
        let get_response = this._get(endpoint);
        let match = get_response.body.match(/name="csrfmiddlewaretoken" value="([^"]+)"/);
        return match ? match[1] : '';
    }

    _post(endpoint, data, need_csrf=true) {
        let url = `${this.base_url}${endpoint}`;
        let requestData = need_csrf ? {...data, csrfmiddlewaretoken: this.getCsrfToken(endpoint)} : data;
        return this.session.post(url, requestData, { redirects: 0 })
    }

    
}